src/Substance.hs

```diff
@@ -4,535 +4,450 @@
+               Left _ -> case checkAndGet p (operations varEnv) pos of
+                 Right o -> checkVarOperation varEnv args o
+                 Left err -> varEnv { errors = (errors varEnv) ++ err }
+
+checkVarPred :: VarEnv -> [PredArg] -> Predicate1 -> VarEnv
+checkVarPred varEnv args (Predicate1 name yls kls tls _) =
+             let exprArgs      = map isVarPredicate args
+                 errAndTypesLs = map (checkExpression varEnv) exprArgs
+                 errls         = map (\(err1, t1) -> err1) errAndTypesLs
+                 err           = foldl (\err1 err2 -> err1 ++ err2) "" errls
+                 argTypes      = map (\(err1, t1) -> t1) errAndTypesLs
+             in if (foldl (\b at1 -> b && isJust at1) True argTypes)
+                then let argTypes2 = map (\a -> KT (fromJust a)) argTypes
+                         tls2      = map (\a -> KT a) tls
+                         sigma     = subst varEnv M.empty argTypes2 tls2
+                     in if (sigma == M.empty)
+                        then varEnv { errors = (errors varEnv) ++ err } -- err should be empty str
+                        else varEnv { errors = (errors varEnv) ++ err } -- err should be empty str
+                else
+                 varEnv { errors = (errors varEnv) ++ err}
+
+checkVarOperation :: VarEnv -> [PredArg] -> Operation -> VarEnv
+checkVarOperation varEnv args (Operation name yls kls tls _) =
+                  let exprArgs      = map isVarPredicate args
+                      errAndTypesLs = map (checkExpression varEnv) exprArgs
+                      errls         = map (\(err1, t1) -> err1) errAndTypesLs
+                      err           = foldl (\err1 err2 -> err1 ++ err2) "" errls
+                      argTypes      = map (\(err1,t1) -> t1) errAndTypesLs
+                  in if (foldl (\b at1 -> b && isJust at1) True argTypes)
+                     then let argTypes2 = map (\a -> KT (fromJust a)) argTypes
+                              tls2      = map (\a -> KT a) tls
+                              sigma     = subst varEnv M.empty argTypes2 tls2
+                          in if (sigma == M.empty)
+                             then varEnv { errors = (errors varEnv) ++ err } -- err should be empty str
+                             else varEnv { errors = (errors varEnv) ++ err } -- err should be empty str
+                     else
+                      varEnv { errors = (errors varEnv) ++ err}
+
+isVarPredicate :: PredArg -> Expr
+isVarPredicate (PP p) = error "Mixed predicate types!"
+isVarPredicate (PE p) = p
+
+isRecursedPredicate :: PredArg -> Predicate
+isRecursedPredicate (PP p) = p
+isRecursedPredicate (PE p) = error "Mixed predicate types!"
+
+checkRecursePred :: VarEnv -> [PredArg] -> VarEnv
+checkRecursePred varEnv args = let predArgs = map isRecursedPredicate args
+                               in foldl checkPredicate varEnv predArgs
+
+checkExpression :: VarEnv -> Expr -> (String, Maybe T)
```

Comment:
to clarify the type signature, I would do `type Error = String` and replace most of these `Strings` with errors